File: app/src/main/java/com/example/energo_monitoring/ui/login/LoginViewModel.java
package com.example.energo_monitoring.ui.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import android.util.Patterns;

import com.example.energo_monitoring.R;
import com.example.energo_monitoring.model.api.AuthorizeBody;
import com.example.energo_monitoring.model.api.AuthorizeResponse;
import com.example.energo_monitoring.presenter.ServerService;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginViewModel extends ViewModel {

    private MutableLiveData<LoginFormState> loginFormState = new MutableLiveData<>();
    private MutableLiveData<AuthorizeResponse> loginResult = new MutableLiveData<>();

    LiveData<LoginFormState> getLoginFormState() {
        return loginFormState;
    }

    LiveData<AuthorizeResponse> getLoginResult() {
        return loginResult;
    }

    public void login(String username, String password) {
        // can be launched in a separate asynchronous job
        ServerService.getService().authorize(new AuthorizeBody(username, password)).enqueue(
                new Callback<AuthorizeResponse>() {
            @Override
            public void onResponse(Call<AuthorizeResponse> call, Response<AuthorizeResponse> response) {
                loginResult.setValue(response.body());
            }

            @Override
            public void onFailure(Call<AuthorizeResponse> call, Throwable t) {
                loginResult.setValue(new AuthorizeResponse(false, t.toString()));
            }
        });
    }

    public void loginDataChanged(String username, String password) {
        if (!isUserNameValid(username)) {
            loginFormState.setValue(new LoginFormState(R.string.invalid_username, null));
        } else if (!isPasswordValid(password)) {
            loginFormState.setValue(new LoginFormState(null, R.string.invalid_password));
        } else {
            loginFormState.setValue(new LoginFormState(true));
        }
    }

    // A placeholder username validation check
    private boolean isUserNameValid(String username) {
        if (username == null) {
            return false;
        }
        if (username.contains("@")) {
            return Patterns.EMAIL_ADDRESS.matcher(username).matches();
        } else {
            return !username.trim().isEmpty();
        }
    }

    // A placeholder password validation check
    private boolean isPasswordValid(String password) {
        return password != null;
    }
}

File: feature-settings/src/main/java/com/example/feature_settings/login/LoginViewModel.kt
package com.example.feature_settings.login

import android.util.Patterns
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.common.AuthorizeBody
import com.example.common.UserData
import com.example.common.api.ServerService
import com.example.feature_settings.AuthorizeResponse
import com.example.feature_settings.R
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class LoginViewModel : ViewModel() {

    class AuthorizeResponse {
        var isSuccess: Boolean
        var user: UserData? = null
        var error = ""

        constructor(success: Boolean, user: UserData?) {
            isSuccess = success
            this.user = user
        }

        constructor(success: Boolean, error: String) {
            isSuccess = success
            this.error = error
        }
    }

    val loginFormState = MutableLiveData<LoginFormState>()
    val loginResult = MutableLiveData<AuthorizeResponse?>()
    fun getLoginFormState(): LiveData<LoginFormState> {
        return loginFormState
    }

    fun getLoginResult(): LiveData<AuthorizeResponse?> {
        return loginResult
    }

    fun login(username: String?, password: String?) {
        // can be launched in a separate asynchronous job
        val auth = object : Callback<AuthorizeResponse> {
            override fun onResponse(call: Call<AuthorizeResponse>, response: Response<AuthorizeResponse>) {
                loginResult.value = response.body()
            }

            override fun onFailure(call: Call<AuthorizeResponse>, t: Throwable) {
                loginResult.value = AuthorizeResponse(false, t.toString())
            }
        }
        ServerService.getService().authorize(AuthorizeBody(username, password)).enqueue(auth)
    }

    fun loginDataChanged(username: String?, password: String?) {
        if (!isUserNameValid(username)) {
            loginFormState.setValue(LoginFormState(R.string.invalid_username, null))
        } else if (!isPasswordValid(password)) {
            loginFormState.setValue(LoginFormState(null, R.string.invalid_password))
        } else {
            loginFormState.setValue(LoginFormState(true))
        }
    }

    // A placeholder username validation check
    private fun isUserNameValid(username: String?): Boolean {
        if (username == null) {
            return false
        }
        return if (username.contains("@")) {
            Patterns.EMAIL_ADDRESS.matcher(username).matches()
        } else {
            !username.trim { it <= ' ' }.isEmpty()
        }
    }

    // A placeholder password validation check
    private fun isPasswordValid(password: String?): Boolean {
        return password != null
    }
}
